BenchmarkSample/benchmark/src/androidTest/java/com/example/benchmark/RecyclerViewBenchmark.kt

```diff
@@ -0,0 +1,88 @@
+import androidx.test.filters.LargeTest
+import com.example.benchmark.ui.MainActivity
+import org.junit.After
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+
+/**
+ * RecyclerView benchmark - scrolls a RecyclerView, and measures the time taken to reveal each item
+ *
+ * You can use this general approach to benchmark the performance of RecyclerView. Some things to be aware of:
+ *
+ * - Benchmark one ItemView type at a time. If you have for example section headers, or other types of item variation,
+ *      it's recommended to use fake adapter data with just a single type of item at a time.
+ * - If you want to benchmark TextView performance, use randomized text. Reusing words between items (such as in this
+ *      simple test) will artificially perform better due to unrealistic caching.
+ * - You won't see the effects of RecyclerView prefetching, or Async text layout with this simple approach. We'll be
+ *      adding more complex RecyclerView examples as time goes on.
+ *
+ * Note that this benchmark measures the sum of multiple potentially expensive stages of displaying an item:
+ * - Attaching an ItemView to RecyclerView
+ * - Detaching an ItemView (scrolling out of viewport) from RecyclerView
+ * - onBindViewHolder
+ * - ItemView layout
+ *
+ * It does *not* measure any of the following work:
+ * - onCreateViewHolder
+ * - RenderThread and GPU Rendering work
+ */
+@LargeTest
+@RunWith(JUnit4::class)
+class RecyclerViewBenchmark {
+
+    @get:Rule
+    val benchmarkRule = BenchmarkRule()
+
+    @get:Rule
+    val activityRule = ActivityScenarioRule(MainActivity::class.java)
+
+    @Before
+    fun setup() {
+        activityRule.scenario.onActivity {
+            // Initialize the Adapter with fake data.
+            // (Submit null first so both are synchronous for simplicity)
+            // ItemViews will be inflated and ready by the next onActivity callback
+            it.adapter.submitList(null)
+            it.adapter.submitList(List(100000) { "List Item $it" })
+        }
+    }
+
+    @After
+    fun teardown() {
+        activityRule.scenario.moveToState(Lifecycle.State.DESTROYED)
+    }
+
+    /**
+     * Measure the cost of scrolling RecyclerView by one ViewHolder.
+     */
+    @UiThreadTest
+    @Test
+    fun simpleScroll() {
+        activityRule.scenario.onActivity {
+            // If RecyclerView has children, the items are attached, bound, and gone through layout. Ready to benchmark.
+            assertTrue("RecyclerView expected to have children",
+                it.recyclerView.childCount > 0)
+
+            benchmarkRule.measureRepeated {
+                // Scroll RecyclerView by one item
+                // this will synchronously execute: attach / detach(old item) / bind / layout
+                it.recyclerView.scrollBy(0, it.recyclerView.getLastChild().height)
```

Comment:
nit; since I don't control the iteration count, this test seems like it might eventually be testing a no-op?

Reply:
It's true that the simple adapter under test will eventually run out of items, but it's initialized with more than enough to scroll through for the benchmark. 100,000 iterations is a lot!

We could submit a bigger list in setUp or rewrite it with an adapter like below - but I think the only way to get a truly infinite list would be to use a PagedList, but that's probably more overhead that would do more harm than good here?
```
private class TrivialAdapter : RecyclerView.Adapter<TrivialViewHolder>() {
    var disableReuse = false
    var createInCode = false

    var inflater: (ViewGroup) -> View = {
        LayoutInflater.from(it.context).inflate(
            R.layout.item_view, it, false)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrivialViewHolder {
        return TrivialViewHolder(inflater(parent))
    }

    override fun onBindViewHolder(holder: TrivialViewHolder, position: Int) {
        val color = position % 256
        holder.itemView.setBackgroundColor(Color.rgb(color, color, color))
    }

    override fun getItemViewType(position: Int): Int {
        return if (disableReuse) position else 0
    }

    override fun getItemCount() = Integer.MAX_VALUE
}
```